internal/expr/query.go

```diff
@@ -61,61 +63,48 @@ func (pe *PreparedExpr) Query(args ...any) (ce *QueryExpr, err error) {
 		if !inQuery[t] {
 			// Check if we have a type with the same name from a different package.
 			for _, typeMember := range pe.inputs {
-				if t.Name() == typeMember.outerType().Name() {
-					return nil, fmt.Errorf("parameter with type %q missing, have type with same name: %q", typeMember.outerType().String(), t.String())
+				if t.Name() == typeMember.OuterType().Name() {
+					return nil, fmt.Errorf("parameter with type %q missing, have type with same name: %q", typeMember.OuterType().String(), t.String())
 				}
 			}
 			return nil, fmt.Errorf("%s not referenced in query", t.Name())
 		}
 	}
 
-	// Query parameteres.
-	qargs := []any{}
+	// Query parameters.
+	var qargs []any
 	for i, typeMember := range pe.inputs {
-		outerType := typeMember.outerType()
+		outerType := typeMember.OuterType()
 		v, ok := typeValue[outerType]
 		if !ok {
 			return nil, typeMissingError(outerType.Name(), typeNames)
 		}
-		var val reflect.Value
-		switch tm := typeMember.(type) {
-		case *structField:
-			val = v.Field(tm.index)
-		case *mapKey:
-			val = v.MapIndex(reflect.ValueOf(tm.name))
-			if val.Kind() == reflect.Invalid {
-				return nil, fmt.Errorf(`map %q does not contain key %q`, outerType.Name(), tm.name)
-			}
+
+		val, err := typeMember.ValueForOuter(v)
```

Comment:
This is one key part of the cleanup. If we have types with a unifying indirection, we should use that indirection to limit the knowledge required by other types.